Add forwarding iptables rules from loopback to eth0

Nino-K opened the pull request:
This allows the traffic that is destined for the loopback interface to also get picked up by the tap device `eth0`. This is to fix the scenario where the published port is bound to `127.0.0.1` e.g `docker run -d -p 127.0.0.1:1041:80 nginx`. In this scenario, there would be no accessible path to the container through `host-switch.exe` since the `vm-switch` is configured to only pick up traffic from `eth0`.

To test:
Run Rancher Desktop as non-admin or admin.
and Then:
```
docker run -d -p 127.0.0.1:1041:80 nginx
```
from windows host:
```
curl.exe localhost:1401
```
should succeed. 

Related to: https://github.com/rancher-sandbox/rancher-desktop/issues/4296

Nino-K committed:
Add forwarding iptables rule from lo to eth0

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

mook-as commented on pkg/config/config.go:
```diff
@@ -0,0 +1,148 @@
+)
+
+const (
+	DefaultSubnet      = "192.168.127.0/24"
+	TapDeviceMacAddr   = "5a:94:ef:e4:0c:ee"
```
Since these are now exported, they should have comments.

mook-as commented on cmd/vm/switch_linux.go:
```diff
@@ -118,14 +121,17 @@ func run(ctx context.Context, cancel context.CancelFunc, connFile io.ReadWriteCl
 
-	if err := linkUp(tapIface, defaultMacAddr); err != nil {
-		logrus.Fatalf("setting mac address [%s] for %s tap device failed: %s", defaultMacAddr, tapIface, err)
+	if err := linkUp(tapIface, config.TapDeviceMacAddr); err != nil {
```
Would it make sense to allow this to be overridden too?

mook-as reviewed (changes requested):
This needs to be documented somewhere (in a markdown file in a repo, I don't mean the docs site).

Nino-K commented:
> This needs to be documented somewhere (in a markdown file in a repo, I don't mean the docs site).

Sure, I will add documents in the README.

Nino-K committed:
Make tap mac address configurable

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

mook-as commented on README.md:
```diff
@@ -50,7 +50,7 @@ portForwarding["Port Forwarding"]
 
 ## vm-switch
-Once the `vm-switch` process starts in the new namespace, it creates a tap (`eth0`) and a `lo` device. The Kernel then forwards all the raw Ethernet frames to the tap device. The tap device forwards the frames over [vsock](https://wiki.qemu.org/Features/VirtioVsock) to the host. The process on the host (`host-switch.exe`) decapsulates the frames, and since it maintains both internal (`vm-switch`  ...
+Once the `vm-switch` process starts in the new namespace, it creates a tap (`eth0`) and a `lo` device. In addition, it creates a series of iptables rules to allow the traffic that is destined for `lo` interface to get picked up by the tap device `eth0`. The Kernel then forwards all the raw Ethernet frames to the tap device. The tap device forwards the frames over [vsock](https://wiki.qemu.org/Fea ...
```
https://github.com/rancher-sandbox/rancher-desktop-networking/blob/76945a86232f720e9936682ee0fb9047d85b6c3c/cmd/vm/switch_linux.go#L158-L162

The rules sound like they're doing the _opposite_ of what's described here? That is, I parse the README changes to mean:
1. Within the network namespace, I run `curl http://localhost:12345`
2. The iptables rules redirects things, as if I ran `curl http://<eth0>:12345`.

Nino-K replied:
Yes, I will reword it.

mook-as commented on cmd/network/setup_linux.go:
```diff
@@ -26,27 +26,37 @@ import (
+	flag.StringVar(&subnet, "subnet", config.DefaultSubnet,
+		fmt.Sprintf("Subnet range with CIDR suffix that is associated to the tap interface, e,g: %s", config.DefaultSubnet))
+	flag.StringVar(&tapDeviceMacAddr, "tap-mac-address", config.TapDeviceMacAddr,
+		"48 bits mac address that is associated to the tap interface")
```
```suggestion
		"MAC address that is associated to the tap interface")
```
MAC addresses are implied to be 48 bits (also, that's an acronym, "Media Access Control" — See IEEE 802 _overview and architecture_ section 4).

Nino-K replied:
> also, that's an acronym, "Media Access Control" — See IEEE 802 _overview and architecture_ section 4

So my assumption about it being the address of Mac and Cheese was wrong all this time? 🤔

mook-as commented on cmd/vm/switch_linux.go:
```diff
@@ -34,24 +34,30 @@ import (
 	flag.BoolVar(&debug, "debug", true, "enable debug flag")
 	flag.StringVar(&tapIface, "tap-interface", defaultTapDevice, "tap interface name, eg. eth0, eth1")
+	flag.StringVar(&tapDeviceMacAddr, "tap-mac-address", config.TapDeviceMacAddr,
+		"48 bits mac address that is associated to the tap interface")
```
See above.

mook-as commented on cmd/vm/switch_linux.go:
```diff
@@ -140,6 +149,34 @@ func run(ctx context.Context, cancel context.CancelFunc, connFile io.ReadWriteCl
+	}
+	tapDeviceIP := config.TapDeviceIP(ip.To4())
+	rules := map[string][]string{
+		"PREROUTING":  {"-t", "nat", "-A", "PREROUTING", "-d", tapDeviceIP, "-j", "DNAT", "--to-destination", "127.0.0.1"},
```
Does that mean if I have two containers, one listening on `127.0.0.1:80`, and one listening on `192.168.127.2:80` (the tap device IP, in case I got that wrong), we'll _always_ send traffic to the former? In fact, if we only have the latter, it'll never receive traffic despite nothing is listening on the former?

(The only scenario I can think of where this matters is actually two containers, `127.0.0.1:80` plus `0.0.0.0:80`.)

Nino-K replied:
I don't think this is a real scenario and the userland proxy (`host-switch.exe`) does have a guard for this and will error out. Consider the following:
1) A user uses port `7000` to deploy a container attached to `127.0.0.1`
e.g:
```
docker run -d -p  127.0.0.1:7000:80 nginx
```
and now you will not be able to deploy a second container on the same port `7000` attached to `0.0.0.0`, the `host-switch.exe` will error out letting the user know that the port is occupied. 

2) A user will attempt to use port `7000` to deploy a container attached to `0.0.0.0`, but the user is not able to re-use that port on `127.0.0.1` since the port is already occupied using `INADDR_ANY` and that includes all local network interfaces (including loopback).
e.g:
```
docker run -d -p  7000:80 nginx
```
3) A user will attempt something clever where they will bind directly to an internal IP address.
e.g:
```
docker run -d -p  192.168.127.2:7000:80 nginx
```
While this a legitimate use case, however, the above port is being forwarded on the host as `127.0.0.1:7000`, 
therefore any attempts to re-use port `7000` on `0.0.0.0` or `127.0.0.1` will prevent the user from successfully running the command and will be prevented by an error from `host-switch.exe`

Nino-K committed:
Use long format options for a better readability

- Also, remove the OUTPUT chain rule since we do not want
to rewrite the destination IP address of any packets generated
by the local system and destined for 192.168.127.2 to 127.0.0.1.

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

Nino-K committed:
Remove the iptable rules upon shutdown

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

Nino-K committed:
Reword the README iptables

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

mook-as reviewed (approved):
🧀
